File: src/core/generation/confidence.py
from typing import Optional

from src.core.schemas import ClaimAudit, Confidence, GateDecision, GroundedAnswer, ThresholdDecision


def compute_confidence(
    gate: GateDecision,
    threshold: Optional[ThresholdDecision],
    claim_audit: Optional[ClaimAudit],
    raw_answer: Optional[GroundedAnswer],
) -> Confidence:
    if (threshold is not None and threshold.action == "HALT") or (raw_answer is not None and raw_answer.insufficient_evidence):
        return "Insufficient Evidence"

    has_unsupported_claim = claim_audit is not None and claim_audit.unsupported_rate > 0
    has_failed_quote = claim_audit is not None and any(not qc.verified for qc in claim_audit.quote_checks)
    if (threshold is not None and threshold.action == "DOWNGRADE") or has_unsupported_claim or has_failed_quote:
        return "Low"

    if gate.verdict == "CAUTION":
        return "Medium"

    return "High"

File: frontend/src/lib/confidence.ts
import { BadgeCheck, CircleAlert, CircleX, ShieldAlert, ShieldCheck, ShieldX, TriangleAlert } from "lucide-react";
import type { Confidence, GateDecision, ThresholdDecision } from "@/types/api";

export interface StatusVisual {
  label: string;
  icon: typeof ShieldCheck;
  badgeClass: string;
  textClass: string;
}

export const gateVisuals: Record<GateDecision["verdict"], StatusVisual> = {
  ALLOW: {
    label: "Allowed",
    icon: ShieldCheck,
    badgeClass: "bg-success/10 text-success border-success/30",
    textClass: "text-success",
  },
  CAUTION: {
    label: "Caution",
    icon: ShieldAlert,
    badgeClass: "bg-warning/10 text-warning border-warning/30",
    textClass: "text-warning",
  },
  REFUSE: {
    label: "Refused",
    icon: ShieldX,
    badgeClass: "bg-destructive/10 text-destructive border-destructive/30",
    textClass: "text-destructive",
  },
};

export const confidenceVisuals: Record<Confidence, StatusVisual> = {
  High: {
    label: "High",
    icon: BadgeCheck,
    badgeClass: "bg-success/10 text-success border-success/30",
    textClass: "text-success",
  },
  Medium: {
    label: "Medium",
    icon: CircleAlert,
    badgeClass: "bg-warning/10 text-warning border-warning/30",
    textClass: "text-warning",
  },
  Low: {
    label: "Low",
    icon: TriangleAlert,
    badgeClass: "bg-warning/10 text-warning border-warning/30",
    textClass: "text-warning",
  },
  "Insufficient Evidence": {
    label: "Insufficient Evidence",
    icon: CircleX,
    badgeClass: "bg-destructive/10 text-destructive border-destructive/30",
    textClass: "text-destructive",
  },
};

export const thresholdVisuals: Record<ThresholdDecision["action"], StatusVisual> = {
  PROCEED: gateVisuals.ALLOW,
  DOWNGRADE: {
    label: "Downgraded",
    icon: TriangleAlert,
    badgeClass: "bg-warning/10 text-warning border-warning/30",
    textClass: "text-warning",
  },
  HALT: {
    label: "Halted",
    icon: ShieldX,
    badgeClass: "bg-destructive/10 text-destructive border-destructive/30",
    textClass: "text-destructive",
  },
};
